revocation/crl/cache/bundle.go

```diff
@@ -0,0 +1,199 @@
+	BaseCRL  FileInfo  `json:"base.crl"`
+	CreateAt time.Time `json:"createAt"`
+}
+
+// FileInfo stores the URL and creation time of the file
+type FileInfo struct {
+	URL string `json:"url"`
+}
+
+// NewBundle creates a new CRL store with tarball format
+func NewBundle(baseCRL *x509.RevocationList, url string) (*Bundle, error) {
+	return &Bundle{
+		BaseCRL: baseCRL,
+		Metadata: Metadata{
+			BaseCRL: FileInfo{
+				URL: url,
+			},
+			CreateAt: time.Now(),
+		},
+	}, nil
+}
+
+// ParseBundleFromTarball parses the CRL blob from a tarball
+//
+// The tarball should contain two files:
+// - base.crl: the base CRL in DER format
+// - metadata.json: the metadata of the CRL
+//
+// example of metadata.json:
+//
+//	{
+//	  "base.crl": {
+//	    "url": "https://example.com/base.crl"
+//	  },
+//	  "createAt": "2024-07-20T00:00:00Z"
+//	}
+func ParseBundleFromTarball(data io.Reader) (*Bundle, error) {
+	bundle := &Bundle{}
+
+	// parse the tarball
+	tar := tar.NewReader(data)
+	for {
+		header, err := tar.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, &BrokenFileError{
+				Err: fmt.Errorf("failed to read tarball: %w", err),
+			}
+		}
+
+		switch header.Name {
+		case PathBaseCRL:
+			// parse base.crl
+			data, err := io.ReadAll(tar)
+			if err != nil {
+				return nil, err
+			}
+
+			var baseCRL *x509.RevocationList
+			baseCRL, err = x509.ParseRevocationList(data)
+			if err != nil {
+				return nil, &BrokenFileError{
+					Err: fmt.Errorf("failed to parse base CRL from tarball: %w", err),
+				}
+			}
+			bundle.BaseCRL = baseCRL
+		case PathMetadata:
+			// parse metadata
+			var metadata Metadata
+			if err := json.NewDecoder(tar).Decode(&metadata); err != nil {
+				return nil, &BrokenFileError{
+					Err: fmt.Errorf("failed to parse CRL metadata from tarball: %w", err),
+				}
+			}
+			bundle.Metadata = metadata
+		default:
+			return nil, &BrokenFileError{
+				Err: fmt.Errorf("unexpected file in CRL tarball: %s", header.Name),
+			}
+		}
+	}
+	if err := bundle.validate(); err != nil {
+		return nil, err
+	}
+
+	return bundle, nil
+}
+
+func (b *Bundle) validate() error {
+	if b.BaseCRL == nil {
+		return errors.New("base CRL is missing")
+	}
+	if b.Metadata.BaseCRL.URL == "" {
+		return errors.New("base CRL URL is missing")
+	}
+	if b.Metadata.CreateAt.IsZero() {
+		return errors.New("base CRL creation time is missing")
+	}
+	return nil
+}
+
+// SaveAsTar saves the CRL blob as a tarball, including the base CRL and
+// metadata
+//
+// The tarball should contain two files:
+// - base.crl: the base CRL in DER format
+// - metadata.json: the metadata of the CRL
+//
+// example of metadata.json:
+//
+//	{
+//	  "base.crl": {
+//	    "url": "https://example.com/base.crl"
+//	  },
+//	  "createAt": "2024-06-30T00:00:00Z"
+//	}
+func (b *Bundle) SaveAsTarball(w io.Writer) (err error) {
+	if err := b.validate(); err != nil {
+		return err
+	}
+
+	tarWriter := tar.NewWriter(w)
+	defer tarWriter.Close()
```

Comment:
The error of `Close()` should not be ignored.

Reply:
Updated.